File: ios/MendixNative/NativeOtaModule/OtaHelpers.m
#import "OtaHelpers.h"
#import "OtaConstants.h"
#import "NativeFsModule.h"

@implementation OtaHelpers



#pragma mark - Implementation

+ (NSString *)resolveAppVersion {
    return [[[NSBundle mainBundle].infoDictionary
        objectForKey:@"CFBundleShortVersionString"]
        stringByAppendingString:
            [@"-"
                stringByAppendingString:[[NSBundle mainBundle].infoDictionary
                                            objectForKey:@"CFBundleVersion"]]];
}

+ (NSString *)getOtaDir {
    return [[NSSearchPathForDirectoriesInDomains(
        NSApplicationSupportDirectory, NSUserDomainMask, YES) firstObject]
        stringByAppendingString:[@"/" stringByAppendingString:[[[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleIdentifier"] stringByAppendingString:[@"/" stringByAppendingString:OTA_DIR_NAME]]]];
}

+ (NSString *)getOtaManifestFilepath {
    return
        [OtaHelpers resolveAbsolutePathRelativeToOtaDir:
                             [@"/" stringByAppendingString:MANIFEST_FILE_NAME]];
}

+ (NSString *)resolveAbsolutePathRelativeToOtaDir:(NSString *)path {
    return [[OtaHelpers getOtaDir] stringByAppendingString:path];
}

+ (NSDictionary *)readManifestAsDictionary {
		NSData *contents = [NSData dataWithContentsOfFile:[OtaHelpers getOtaManifestFilepath] options:nil error:nil];
  	if (contents == nil) {
  		return nil;
  	}
  	NSDictionary *jsonOutput = [NSJSONSerialization JSONObjectWithData:contents options:kNilOptions error:nil];
  	if (jsonOutput == nil) {
   	 	return nil;
  	}
   	return jsonOutput;
}

+ (NSDictionary *) getNativeDependencies {
    return [NativeFsModule readJson:[[NSBundle mainBundle] pathForResource:@"native_dependencies" ofType:@"json"] error:nil] ?: @{};
}

@end

File: ios/Modules/NativeOtaModule/OtaHelpers.swift
import Foundation

class OtaHelpers: NSObject {
    
    private static func version() -> String {
        let info = Bundle.main.infoDictionary
        return info?["CFBundleVersion"] as? String ?? ""
    }
    
    private static func shortVersion() -> String {
        let info = Bundle.main.infoDictionary
        return info?["CFBundleShortVersionString"] as? String ?? ""
    }
    
    private static func identifier() -> String {
        let info = Bundle.main.infoDictionary
        return info?["CFBundleIdentifier"] as? String ?? ""
    }
    
    static func resolveAppVersion() -> String {
        return "\(shortVersion())-\(version())"
    }
    
    static func getOtaDir() -> String {
        let supportDirectory = NSSearchPathForDirectoriesInDomains(.applicationSupportDirectory, .userDomainMask, true).first ?? ""
        return "\(supportDirectory)/\(identifier())/\(OTA_DIR_NAME)"
    }
    
    static func getOtaManifestFilepath() -> String {
        return resolveAbsolutePathRelativeToOtaDir("/\(MANIFEST_FILE_NAME)")
    }
    
    static func resolveAbsolutePathRelativeToOtaDir(_ path: String) -> String {
        return "\(getOtaDir())\(path)"
    }
    
    static func readManifestAsDictionary() -> [String: Any]? {
        let manifestPath = getOtaManifestFilepath()
        
        guard let contents = NSData(contentsOfFile: manifestPath) else {
            return nil
        }
        
        do {
            let jsonOutput = try JSONSerialization.jsonObject(with: contents as Data, options: [])
            return jsonOutput as? [String: Any]
        } catch {
            return nil
        }
    }
    
    static func getNativeDependencies() -> [String: Any] {
        guard let path = Bundle.main.path(forResource: "native_dependencies", ofType: "json") else {
            return [:]
        }
        
        return NativeFsModule.readJson(path, error: nil) ?? [:]
    }
}

//Checked
